ChiModules/CFEMDiffusion/cfem_diffusion_solver.cc

```diff
@@ -0,0 +1,337 @@
+              CallLua_iXYZFunction(L, "Sigma_a",imat,qp_data.QPointXYZ(qp))  *
+              qp_data.ShapeValue(i, qp) * qp_data.ShapeValue(j, qp)
+            )
+            *
+            qp_data.JxW(qp);
+        }//for qp
+        Acell[i][j] = entry_aij;
+      }//for j
+      for (size_t qp : qp_data.QuadraturePointIndices())
+        cell_rhs[i] += CallLua_iXYZFunction(L,"Q_ext",imat,qp_data.QPointXYZ(qp)) * qp_data.ShapeValue(i, qp) * qp_data.JxW(qp);
+    }//for i
+ 
+    //======================= Flag nodes for being on a boundary
+    std::vector<int> dirichlet_count(num_nodes, 0);
+    std::vector<double> dirichlet_value(num_nodes, 0.0);
+
+    const size_t num_faces = cell.faces.size();
+    for (size_t f=0; f<num_faces; ++f)
+    {
+      const auto& face = cell.faces[f];
+      // not a boundary face
+	    if (face.has_neighbor) continue; 
+	  
+      const auto& bndry = boundaries[face.neighbor_id];
+
+      // Robin boundary
+      if (bndry.type == BoundaryType::Robin)
+      { 
+        const auto  qp_face_data = cell_mapping.MakeFaceQuadraturePointData( f );
+        const size_t num_face_nodes = face.vertex_ids.size();
+
+        const auto& aval = bndry.values[0];
+        const auto& bval = bndry.values[1];
+        const auto& fval = bndry.values[2];
+
+        chi::log.Log() << "Boundary  set as Robin with a,b,f = ("
+                    << aval << ","
+                    << bval << ","
+                    << fval << ") ";
+        // true Robin when a!=0, otherwise, it is a Neumann:
+        // Assert if b=0
+        if (std::fabs(bval) < 1e-8)
+          throw std::logic_error("if b=0, this is a Dirichlet BC, not a Robin BC");
+        
+        // loop over nodes of that face
+        for (size_t fi=0; fi<num_face_nodes; ++fi)
+        {
+          const uint i = cell_mapping.MapFaceNode(f,fi);
+            
+          double entry_rhsi = 0.0;
+          for (size_t qp : qp_face_data.QuadraturePointIndices() )
+            entry_rhsi +=  qp_face_data.ShapeValue(i, qp) * qp_face_data.JxW(qp);
+          cell_rhs[i] +=  fval / bval * entry_rhsi;
+            
+          // only do this part if true Robin (i.e., a!=0)
+          if (std::fabs(aval) > 1.0e-8)
+          {
+            for (size_t fj=0; fj<num_face_nodes; ++fj)
+            {
+              const uint j = cell_mapping.MapFaceNode(f,fj);
+          
+              double entry_aij = 0.0;
+              for (size_t qp : qp_face_data.QuadraturePointIndices())
+                entry_aij +=  qp_face_data.ShapeValue(i, qp) *qp_face_data.ShapeValue(j, qp)
+                                * qp_face_data.JxW(qp);
+              Acell[i][j] += aval / bval * entry_aij;
+            }//for fj
+          }//end true Robin
+        }//for fi
+      }//if Robin
+	  
+	    // Dirichlet boundary
+	    if (bndry.type == BoundaryType::Dirichlet)
+      {
+ 		    const size_t num_face_nodes = face.vertex_ids.size();
+
+        const auto& boundary_value = bndry.values[0];
```

Comment:
This can have a default value associated with the boundary conditions but for MMS we will have to get creative. We can talk more on this later.

Reply:
We could define a basic option called "DirichletFunc" and default it to empty `""`. This way the boundary value will be used like it is here unless `not dirichlet_func_name.empty()` then we call the lua function with the node XYZ position.